[Not ready to merge yet] XD-3319 First cut at providing an hdfs-sink module

trisberg opened the pull request:
For  testing purposes.

This works with Pivotal HD 3.0, need to be able to swap that out for other distros some day ...

And we need to add some tests ...

trisberg committed:
XD-3319 First cut at providing an hdfs-sink module

ericbottard commented on hdfs-sink/pom.xml:
```diff
@@ -0,0 +1,60 @@
+	</parent>
+
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Hopefully https://github.com/spring-cloud/spring-cloud-stream-modules/pull/15 will be merged first, will need to revisit this pom then

ericbottard commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/DataStoreWriterFactoryBean.java:
```diff
@@ -0,0 +1,187 @@
+
+	@Override
+	public void destroy() throws Exception {
+		storeWriter = null;
```
What is the purpose of this? The bean is getting destroyed anyway

ericbottard commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSinkApplication.java:
```diff
@@ -0,0 +1,35 @@
+ * @author Thomas Risberg
+ */
+@SpringBootApplication
+@ComponentScan(basePackageClasses= HdfsSink.class)
```
Redundant with `@SpringBootApplication`

ericbottard commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSinkProperties.java:
```diff
@@ -0,0 +1,215 @@
+	}
+
+
+	@Pattern(regexp = "(?i)(GZIP|SNAPPY|BZIP2|LZO|SLZO)",
```
Case use an enum here

ericbottard commented on hdfs-sink/src/main/resources/application.yml:
```diff
@@ -0,0 +1,21 @@
+    stream:
+      bindings:
+        input: ticktock
+      module:
```
~~Those would rather go as defaults in the fields~~
I think we'll want to remove those, actually

ericbottard commented on hdfs-sink/src/main/resources/application.yml:
```diff
@@ -0,0 +1,21 @@
+  cloud:
+    stream:
+      bindings:
+        input: ticktock
```
This will need to change

ericbottard commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSink.java:
```diff
@@ -0,0 +1,51 @@
+	@ServiceActivator(inputChannel=Sink.INPUT)
+	public void hdfsSink(Object payload) {
+		try {
+			dataStoreWriter.write(payload.toString());
```
Do we need to guard against `null` payload here?

trisberg committed:
XD-3319 Second cut at providing an hdfs-sink module

- updated according to the current thinking of module configurations

- built for Pivotal HD 3.0

dturanski commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSink.java:
```diff
@@ -0,0 +1,53 @@
+@EnableModule(Sink.class)
+public class HdfsSink {
+
+	DataStoreWriter<String> dataStoreWriter;
```
Why not just @Autowire this?

trisberg replied:
Did you see the setter?

dturanski replied:
Yes, no setter needed unless I'm missing something

dturanski commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSink.java:
```diff
@@ -0,0 +1,53 @@
+
+	@ServiceActivator(inputChannel=Sink.INPUT)
+	public void hdfsSink(Object payload) {
+		if (payload != null) {
```
SI will not send null payload

trisberg replied:
Hmm, added that on suggestion from @ericbottard

dturanski commented on hdfs-sink/src/main/java/org/springframework/cloud/stream/module/hdfs/sink/HdfsSinkProperties.java:
```diff
@@ -0,0 +1,217 @@
+ *
+ * @author Thomas Risberg
+ */
+@ConfigurationProperties(prefix = "spring.cloud.stream.module.hdfs.sink")
```
We don't use prefix for modules

trisberg replied:
Copied that from somewhere, did we change this recently?

dturanski replied:
IDK - I thought we wanted simple names as currently in XD, however this may be useful for composed modules. I'll check the others.

dturanski replied:
All the others use simple names

dturanski commented on pom.xml:
```diff
@@ -37,6 +37,7 @@
 		<module>counter-sink</module>
 		<module>log-sink</module>
 		<module>redis-sink</module>
+		<module>hdfs-sink</module>
```
let's keep in alphabetical order

trisberg committed:
XD-3319 Second cut at providing an hdfs-sink module

- removing prefix for config options

- addressing review comments
